lib/ansible/modules/cloud/docker/docker_swarm_facts.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/python
+#
+# (c) 2019 Piotr Wojciechowski <piotr@it-playground.pl>
+# (c) Thierry Bouvet (@tbouvet)
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: docker_swarm_facts
+
+short_description: Retrieves facts about Docker Swarm cluster.
+
+description:
+  - Retrieves facts about a Docker Swarm if run on Swarm Manager host.
+  - Retrieves role of node on which it is run.
+
+version_added: "2.8"
+
+author:
+    - Piotr Wojciechowski (@wojciechowskipiotr)
+    - Thierry Bouvet (@tbouvet)
+
+options:
+  list:
+    description:
+      - Returns lists of components.
+      - If C(nodes) the return list of registered nodes (the only supported option now).
+      - If not present then return Swarm facts.
+    choices:
+        - nodes
+    required: false
+  output:
+    description:
+      - If C(short) and I(list) is C(nodes) then returns list of registered nodes names.
+      - If C(long) and I(list) is C(nodes) then returns list as in output of `docker nodes ls`.
+      - Ignored if I(list) is not present.
```

Comment:
How about having only one option, `list`, with values `nodes` and `node-hostnames`?

Reply:
You are right, it is a better approach. But we need a better name than `node-hostnames`. It is misleading because the current `long` output returns five to seven attributes for each node depending on its role and state. Maybe `nodes-info` or `nodes-basics`?

Reply:
`node-hostnames` (for me) would be the list of hostnames, while `nodes` would be the list with more infos. Why not return the list of inspect results, though?

How about three modes:
 - `nodes`: list of inspect results
 - `node-hostnames`: list of hostnames
 - `node-summaries`: list of summaries (ID, hostname, status, ...)

Reply:
How about following with the approach from the `docker_host_facts`?

`swarm:` - swarm info 
`nodes:` - lists nodes with basic info like in `docker node ls`
`nodes_filters:` - filtering the query range
`tasks:`- lists tasks
`tasks_filters:` - filtering the query range
`services:` - lists docker services running on swarm
`services_filters:` - filtering the query range
`verbose_output:` - instead if matching the CLI output it will return everything it gets from the API call

This way we will have a unified approach between `docker_host_facts` and `docker_swarm_facts`.

Reply:
I'm a fan of this approach @WojciechowskiPiotr for this situation.

Reply:
I like that idea as well!